sos/plugins/origin.py

```diff
@@ -0,0 +1,132 @@
+# Copyright (C) 2016 Red Hat, Inc., Pep Turro Mauri <pep@redhat.com>
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+
+from sos.plugins import Plugin, RedHatPlugin
+import os.path
+
+
+# This plugin collects static configuration and runtime information
+# about OpenShift Origin based environments, like Atomic Platform or
+# OpenShift Enterprise 3+.
+
+# Some clarifications on naming.
+# OpenShift Origin is the upstream project for OpenShift Enterprise.
+# However, it actually refers to two different code bases:
+#  * Origin M5 and later (https://github.com/openshift/origin)
+#    which is upstream for OpenShift 3.x and later.
+#    This is what this plugin handles
+#  * Origin M4 and earlier (https://github.com/openshift/origin-server)
+#    which is upstream for OpenShift 1.x and 2.x.
+#    This is handled by the plugin in openshift.py
+# Origin M5 is also the code base behind Atomic Platform,
+# not to be confused with Atomic Host (http://projectatomic.io)
+
+# Note that this plugin should be used in conjunction with other plugins
+# in order to capture relevant data: the Kubernetes plugin for the
+# masters, the Docker plugin for the nodes, and also generic
+# plugins (e.g. for /etc/sysconfig entries, network setup etc)
+class OpenShiftOrigin(Plugin):
+    ''' OpenShift Origin '''
+
+    master_base_dir = "/etc/origin/master"
+    node_base_dir = "/etc/origin/node"
+    master_cfg = os.path.join(master_base_dir, "master-config.yaml")
+    node_cfg = os.path.join(node_base_dir, "node-config.yaml")
+    admin_cfg = os.path.join(master_base_dir, "admin.kubeconfig")
+    oc_cmd = "oc --config=%s" % admin_cfg
+
+    files = (master_cfg, node_cfg)
+    profiles = ('openshift',)
+    plugin_name = "origin"
+
+    def is_master(self):
+        '''Determine if we are on a master'''
+        return os.path.exists(self.master_cfg)
+
+    def is_node(self):
+        '''Determine if we are on a node'''
+        return os.path.exists(self.node_cfg)
+
+    def setup(self):
+        if self.is_master():
+            self.add_copy_spec([
+                self.master_cfg,
+                os.path.join(self.master_base_dir, "*.crt"),
+            ])
```

Comment:
The problem here is that masters can also be nodes - and in fact they are on most deployments. Therefore, when ran from a master we need to collect info for master and node, with their respective different base dirs.

I see your point and I appreciate that this seems to have potential for optimization, but I believe that in this case trying to unify this would make things more confusing in the end.